File: Week 3/rust/1.for_loop.rs
/*
 * WEEK 3 - RUST LOOPS & NUMBER THEORY
 * Topic: For Loop Basics
 * File: 1.for_loop.rs
 *
 * CONCEPT:
 *  Rust has only one form of `for`: `for pattern in iterator { body }`.
 *  There is no C-style `for(init; cond; update)`. To iterate a counter,
 *  use a Range like `0..n` (exclusive) or `0..=n` (inclusive).
 *
 * KEY POINTS:
 *  - `0..n`     -> 0, 1, ..., n-1 (Range)
 *  - `0..=n`    -> 0, 1, ..., n
 *  - `(0..n).rev()` -> n-1, ..., 0
 *  - `(0..n).step_by(k)` -> 0, k, 2k, ...
 *  - For containers: `for x in &vec { ... }` borrows; `for x in vec` consumes.
 *  - `.enumerate()` yields (index, value) pairs.
 *  - `.zip()` parallel iteration.
 *
 * SYNTAX:
 *   for i in 0..3 { println!("Inside for loop : {i}"); }
 *
 * DRY RUN:
 *   "Inside for loop : 0", ..., "Inside for loop : 2", "Done"
 *
 * COMPLEXITY: O(n).
 */

fn main() {
    for i in 0..3 {
        println!("Inside for loop : {i}");
    }
    println!("Done");

    println!("\n--- inclusive 1..=5 ---");
    for i in 1..=5 { println!("{i}"); }

    println!("\n--- step_by ---");
    for i in (0..10).step_by(2) { print!("{i} "); }
    println!();

    println!("\n--- reverse ---");
    for i in (0..5).rev() { print!("{i} "); }
    println!();

    println!("\n--- enumerate ---");
    let arr = ["a", "b", "c"];
    for (i, v) in arr.iter().enumerate() {
        println!("  index {i}: {v}");
    }

    println!("\n--- zip ---");
    let a = [1, 2, 3];
    let b = ["x", "y", "z"];
    for (x, y) in a.iter().zip(b.iter()) {
        println!("  pair {x}-{y}");
    }
}

/*
 * NOTES:
 *  - There is NO C-style `for(init; cond; update)`. Always iterate over an Iterator.
 *  - Borrowing matters: `for x in vec` MOVES the vector; `for x in &vec` borrows.
 *  - Iterator combinators (.filter, .map, .take, .collect) replace many loops.
 *  - `loop` is the unconditional infinite loop.
 */

File: Week 3/python/1.for_loop.py
"""
WEEK 3 - PYTHON LOOPS & NUMBER THEORY
Topic: For Loop Basics
File: 1.for_loop.py

CONCEPT:
Python's `for` loop is FUNDAMENTALLY DIFFERENT from Java's. It iterates
over the items of any ITERABLE (list, tuple, str, range, generator, ...).
There is NO `for (init; cond; update)` C-style header.

To imitate Java's `for (int i = 0; i < n; i++)`, use `for i in range(n):`.

KEY POINTS:
 - `range(stop)`            -> 0, 1, ..., stop-1
 - `range(start, stop)`     -> start, ..., stop-1
 - `range(start, stop, step)` -> with custom step (negative for descending)
 - `range` is lazy: it generates values on demand (no list created).
 - `for-else`: the `else:` clause runs after a NORMAL exit (not after break).
 - `enumerate(iterable, start=0)` yields (index, value) pairs.
 - `zip(a, b)` pairs items from two iterables.

SYNTAX:
 for i in range(n): ...
 for i in range(1, 11): ...
 for x in [1, 2, 3]: ...
 for i, v in enumerate(arr): ...

DRY RUN:
 for i in range(3): print("Inside for loop :", i)
   -> 0, 1, 2 (each preceded by "Inside for loop :")
 print("Done")

COMPLEXITY: O(n) — n iterations.
"""


def basic_for_loop() -> None:
    print("--- basic ---")
    for i in range(3):              # 0, 1, 2
        print("Inside for loop :", i)
    print("Done")


def with_explicit_init() -> None:
    """No need to put init outside the for; range handles it."""
    print("\n--- range with start ---")
    for i in range(1, 6):           # 1..5 inclusive
        print(i)


def for_else_demo() -> None:
    """The else block runs ONLY if the loop completes without `break`."""
    print("\n--- for-else ---")
    target = 7
    for x in range(10):
        if x == target:
            print(f"found {target} -> breaking")
            break
    else:
        print("completed without finding")

    # Same loop without break
    for x in range(3):
        print(f"x = {x}")
    else:
        print("no break -> else runs")


def helpful_iter_helpers() -> None:
    """enumerate, zip, reversed -- common idioms."""
    print("\n--- enumerate / zip / reversed ---")
    arr = ["a", "b", "c"]
    for i, v in enumerate(arr):
        print(f"  index {i}: {v}")

    a = [1, 2, 3]
    b = ["x", "y", "z"]
    for x, y in zip(a, b):
        print(f"  pair {x}-{y}")

    for v in reversed(range(5)):
        print(f"  rev {v}")


def main() -> None:
    basic_for_loop()
    with_explicit_init()
    for_else_demo()
    helpful_iter_helpers()


if __name__ == "__main__":
    main()


# NOTES:
# - Java's `for (init; cond; update)` doesn't exist; Python uses `for in iterable`.
# - To replicate Java's index-based loop, use `range(...)`.
# - `range(start, stop, step)` step CAN be negative for descending.
# - `for-else` is unique to Python -- the else runs after a NORMAL completion.
# - Avoid `for i in range(len(arr))`; prefer `for i, v in enumerate(arr)`.
